File: .old_code/Archer-nw2/geometries/sphere.py
import taichi as ti
from vectors import vec3, Ray, normalize

@ti.dataclass
class Sphere:
    """
    A sphere geometry
    """
    position: vec3
    radius: ti.f32

    @ti.func
    def intersect(self, ray: Ray) -> ti.f32:
        """
        Intersects a ray with the sphere
        """
        # Calculate the vector from the sphere's center to the ray's origin
        oc = ray.origin - self.position

        # Calculate the discriminant
        a = ti.math.dot(ray.direction, ray.direction)
        b = 2 * ti.math.dot(oc, ray.direction)
        c = ti.math.dot(oc, oc) - self.radius**2
        discriminant = b**2 - 4 * a * c

        # If the discriminant is negative, there are no real roots
        t1 = -1.0
        t2 = -1.0
        if discriminant >= 0:
            # Calculate the two intersection points
            t1 = (-b - ti.sqrt(discriminant)) / (2 * a)
            t2 = (-b + ti.sqrt(discriminant)) / (2 * a)

        # Return the closest intersection point
        return min(t1, t2)

    @ti.func
    def normal(self, point: vec3) -> vec3:
        """
        Calculates the normal at a point on the sphere
        """
        return normalize(point - self.position)

File: archer/src/geometries/sphere.rs
use bvh::aabb::Aabb;
use nalgebra::{Point3, Vector3};

use crate::geometries::base::Geometry;
use crate::ray::Ray;

pub struct Sphere {
    pub radius: f64,
    pub position: Point3<f64>,
}

impl Geometry for Sphere {
    fn intersect(&self, incoming: &Ray) -> Option<(Point3<f64>, Vector3<f64>)> {
        // Origin
        let oc = incoming.origin - self.position;

        // Coefficients
        let a = incoming.direction.norm_squared();
        let b = 2.0 * incoming.direction.dot(&oc);
        let c = oc.norm_squared() - self.radius * self.radius;
        let discriminant = b * b - 4.0 * a * c;

        if discriminant < 0.0 {
            return None;
        }

        // Calculate the nearest intersection point
        let t = (-b - discriminant.sqrt()) / (2.0 * a);

        // Shadow Acne fix
        if t < f64::EPSILON {
            return None;
        }

        let point = incoming.position_at(t);
        let normal = (point - self.position).normalize();

        Some((point, normal))
    }

    fn aabb(&self) -> Aabb<f64, 3> {
        let half_size = Vector3::new(self.radius, self.radius, self.radius);
        let min = self.position - half_size;
        let max = self.position + half_size;
        Aabb::with_bounds(min, max)
    }
}
